---
title: "locho_study"
date: 2025.01.31
author: 
 - Matt Jackson
 - Dayakar Badri
 - Artemis S. Louyakis
output: html_notebook
---

Load packages used in the analysis
```{r load packages}
Sys.time()

library(tidyr)
library(tidyverse)
library(edgeR)
library(trinotateR)
library(NOISeq)
library(manhattanly)
library(cowplot)
library(wesanderson)
library(vegan)
library(corrplot)
library(ggtern)
library(pheatmap)
# library(EnhancedVolcano) ## load separately - conflict in list
# library(beepr) ## alarm added to code chunks that may take longer than others
```

Create directories for input data and output files
```{r create dirs}
Sys.time()

data_dir <- "data"
plots_dir <- "plots"
tables_dir <- "tables"
noiseq_dir <- "noiseq"

if (!dir.exists(data_dir)) {
  dir.create(data_dir)
}
if (!dir.exists(tables_dir)) {
  dir.create(tables_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
if (!dir.exists(noiseq_dir)) {
  dir.create(noiseq_dir)
}
```

import counts and remove zeroes
```{r import counts data}
Sys.time()

## rsem count estimates
counts <- read.delim("alignments/iso.counts.matrix", row.names = 1)
counts <- counts[rowSums(counts) > 0,]

## normalize by gene length and transformed to counts per million
tpm <- read.delim("alignments/iso.TPM.not_cross_norm", row.names = 1)
tpm <- tpm[rowSums(tpm) > 0,]

## tmm normalized counts
# tmm <- read.delim("alignments/iso.TMM.EXPR.matrix", row.names = 1)
# tmm <- tmm[rowSums(tmm) > 0,]

count_totals <- as.data.frame(colSums(counts))
write.table(count_totals, "tables/count_totals.csv", quote = FALSE, col.names = FALSE)
```

import annotations and mapping files
```{r import annotations}
Sys.time()

## gene lengths were pulled from one of the trinity alignment results files
## awk -F"\t" '{OFS="\t"}{print $1,$3}' alignments/mrna_F-50179058-003S/mrna_F-50179058-003S.isoforms.results > data/isoform_length.txt
# len <- read.delim("data/gene_length.txt")
len <- read.delim("data/isoform_length.txt")

## trinotate annotations cutoff - 10e-3
annots <- read.delim("annotations/trinotate_annotation_report.xls", header = TRUE)

## maps for uniprot, kegg
uniprot_annots <- read.delim("/data/databases/uniprot/uniprotkb_AND_reviewed_true_2025_03_04.tsv", header = TRUE)
## kegg org:gene to uniprot map from kegg ftp
kegg2entry <- read.delim("/data/databases/kegg/ftp.pathway.jp/kegg/genes/links/genes_uniprot.list", 
                         header = FALSE, col.names = c("Kegg", "Entry"))
## kegg org:gene to KO id from kegg ftp
kegg2ko <- read.delim("/data/databases/kegg/ftp.pathway.jp/kegg/genes/links/genes_ko.list", 
                      header = FALSE, col.names = c("Kegg", "ko"))

# ko_map <- read.delim("data/kegg_up_map.txt", header = FALSE) %>% rename(kegg = V1, ko = V2, gene = V3, uniprot = V4)
# khier <- read.delim("data/khier_ko.txt", header = FALSE) %>% rename(level1 = V1, level2 = V2, level3 = V3, ko = V4, ko_name = V5) %>% unique()
```

import sample metadata
```{r metadata}
Sys.time()

## metadata is for the companion metagenome
## need to use mapping file to link mt ids to mg metadata
mg_metadata <- read.delim("data/mg_metadata.csv") %>%
  rename_with(~ tolower(gsub(".", "_", .x, fixed = TRUE))) %>%
  mutate(mg = sub("^([^-]*-[^-]*).*", "\\1", metagenomics_sample_number))

mgmt_map <- read.delim("data/bb4_mtmg_map.tsv") %>%
  rename(mt_full = X.mt, mg_full = mg) %>%
  mutate(mg = sub("^([^-]*-[^-]*).*", "\\1", mg_full),
         mt = sub("^([^-]*-[^-]*).*", "\\1", mt_full))

sample_data <- read.delim("data/locho_105536_sample_data.tsv")
results_data <- read.delim("data/locho_105536_results_data.tsv")

metadata <- merge(mg_metadata, mgmt_map, by = "mg", all = TRUE) %>%
  drop_na(mt)

sd <- sample_data %>%
  select(ck_number, animal_id, sample_type, cp_phase, test_day) %>%
  distinct()

rd <- results_data %>%
  select(ck_number, animal_id, animal_name, tattoo_number, sample_type, matrix,
         cp_phase, phase_day, analysis_method, operating_set, analyte, standard_value, standard_unit) %>%
  unite("unique_id", c(animal_name, tattoo_number, cp_phase, phase_day)) %>%
  distinct()

# rdw <- rd %>%
#   pivot_wider(names_from = c(sample_type, matrix, analysis_method, operating_set, analyte),
#               values_from = c(standard_value, standard_unit),
#               id_cols = )
# 
# metadata %>% 
#   select(mg, mt, diet_name)
# 
# rd |>
#   dplyr::summarise(n = dplyr::n(), .by = c(ck_number, animal_id, unique_id, sample_type, matrix, operating_set,
#   analyte)) |>
#   dplyr::filter(n > 1L)
# 
# results_data %>%
#   filter(ck_number == "B-50179072" & analyte == "Basophils")
```

```{r gene length}
Sys.time()

## aggregate counts by gene id when input is isoform table
counts_lc <- counts_lc %>% 
  mutate(X = gsub("_i[0-9]", "", X)) %>%
  group_by(X) %>%
  summarise(across(everything(), sum)) %>%
  column_to_rownames("X")
## filter gene lengths table to remove any not in counts table
len <- len[match(rownames(counts_lc), len$gene_id, nomatch=0),]

## normalize to gene length - 
counts_lc <- counts_lc %>%
  merge(., len, by.x = 0, by.y = "gene_id") %>% 
  column_to_rownames("Row.names") %>%
  mutate_each(funs(./(length/1000)), starts_with("X")) %>% 
  select(-length)
```

## cluster
```{r clustering}
Sys.time()


```

## plotting quality not needed if running fastqc
```{r plot quality}
Sys.time()

qual <- read.delim("quality/fastx_F-50179071-003S_r1_stats.csv", header = TRUE)

ggplot(qual, aes(x = column, group = column)) +
  theme_bw() +
  geom_boxplot(stat = "identity", color="darkgrey", fill="lightgrey", alpha=0.2, 
    aes(lower = Q1, 
        upper = Q3,
        middle = mean,
        ymin = min,
        ymax = max)) 

ggplot(qual, aes(x = column, y = mean)) +
  theme_bw() +
  geom_line(color = "red") +
  scale_x_continuous(breaks = seq(0, 151, by = 10)) +
  scale_y_continuous(breaks = seq(0, 40, len = 5))


```

## filter blast data by coordinates
```{r filter blast}
Sys.time()

library(GenomicRanges)
library(plyranges)

blast_rslt <- read.table("annotations/blastx.outfmt6", header = FALSE, sep = "\t")

colnames(blast_rslt) <- c("query", "sseqid", "pident", "length", "mismatch", "gapopen", 
                          "qstart", "qend", "sstart", "send", "evalue", "bitscore")

test <- blast_rslt %>% dplyr::select(query, sseqid, qstart, qend, evalue) 

test <- test %>% 
  mutate(i_start = case_when(qstart < qend ~ qstart,
                             qend < qstart ~ qend))

test <- test %>% 
  mutate(i_end = case_when(qstart < qend ~ qend,
                           qend < qstart ~ qstart))

test <- test %>% 
  mutate(strand = case_when(qstart < qend ~ "+", 
                            qend < qstart ~ "-"))

test <- test %>% 
  select(query, sseqid, i_start, i_end, evalue, strand)

colnames(test) <- c("query", "seqnames", "start", "end", "evalue", "strand")

test_irange <- test %>% 
  as_granges()

test_disjoin <- reduce(test_irange,with.revmap = TRUE)

list_revmap <- as.data.frame(mcols(test_disjoin))

filtered_data <- c()
for(i in 1:nrow(list_revmap)){
  filtered_data <- c(filtered_data, (slice(list_revmap, i) %>% 
                                       unlist(use.names=FALSE))[which.min(slice(test,  slice(list_revmap, i) %>%
                                                                                  unlist(use.names=FALSE))$evalue)])
}

best_hits <- slice(test, filtered_data)
```

## create metaphlan table at species level for arby to run through multimedia
```{r prep for multimedia}
Sys.time()

metaphlan <- readxl::read_excel("data/msystems.00452-24-s0004.xlsx")
colnames(metaphlan) <- gsub("-005S|-004S|-003S|-002S", "", colnames(metaphlan))
metaphlan <- metaphlan %>% 
  column_to_rownames("Taxonomy\\Sample ID")

# Define a function to create a table for a given taxonomic level
create_taxa_table <- function(metaphlan, level) {
  level_names <- c("domain", "phylum", "class", "order", "family", "genus", "species")
  level_index <- match(level, level_names)
  
  # Create a list of columns to remove based on the current level
  cols_to_remove <- level_names[(level_index + 1):length(level_names)]
  
  metaphlan %>%
    rownames_to_column() %>%
    separate(col = "rowname", sep = "\\|", into = level_names, fill = "right") %>%
    # Filter out rows where the next level is NA, but the current level is not NA
    filter(if (level_index < length(level_names)) is.na(get(level_names[level_index + 1])) & !is.na(get(level)) else !is.na(get(level))) %>%
    # Select the current level and the data columns
    select(all_of(c(level, setdiff(names(metaphlan), cols_to_remove)))) %>%
    # Set the current level as row names
    column_to_rownames(var = level)
}

# List of taxonomic levels
levels <- c("domain", "phylum", "class", "order", "family", "genus", "species")

# Apply the function to each level and store the results in a list
taxa_tables <- map(levels, ~create_taxa_table(metaphlan, .x))

# Set the names of the list elements to match the taxonomic levels
names(taxa_tables) <- levels

# Now you can access the phylum table using the name
species_table <- taxa_tables[["species"]]
write.csv(species_table, "tables/metaphlan_species.csv", quote = FALSE)
```




TODO: HIGH PRIORITY cluster analysis
TODO: account for repeated measure with diet as main effect and order of diet as repeated measure
TODO: WGCNA 
TODO: RLQ - which transcription factors controlling expression of each gene [Baty et al., 2013](https://pmc.ncbi.nlm.nih.gov/articles/PMC3686578/); [RLQ R tutorial](https://rfrelat.github.io/RLQ.html); [matt's google drive on rlq](https://drive.google.com/drive/folders/1BTs1d1G__SJJWVmdvMHpZJOLJmAZ16GO)

```{r session info}
Sys.time()

sessionInfo()
```


### END ###