---
title: "locho_study"
output: html_notebook
---


```{r load packages}
Sys.time()

library(tidyverse)
```

```{r import data}
Sys.time()

counts <- read.delim("alignments/iso.counts.matrix", row.names = 1)
tpm <- read.delim("alignments/iso.TPM.not_cross_norm", row.names = 1)
# tmm <- read.delim("alignments/iso.TMM.EXPR.matrix", row.names = 1)

colSums(counts)
```


## cluster
```{r clustering}
Sys.time()


```

## plotting quality not needed if running fastqc
```{r plot quality}
Sys.time()

qual <- read.delim("quality/fastx_F-50179071-003S_r1_stats.csv", header = TRUE)

ggplot(qual, aes(x = column, group = column)) +
  theme_bw() +
  geom_boxplot(stat = "identity", color="darkgrey", fill="lightgrey", alpha=0.2, 
    aes(lower = Q1, 
        upper = Q3,
        middle = mean,
        ymin = min,
        ymax = max)) 

ggplot(qual, aes(x = column, y = mean)) +
  theme_bw() +
  geom_line(color = "red") +
  scale_x_continuous(breaks = seq(0, 151, by = 10)) +
  scale_y_continuous(breaks = seq(0, 40, len = 5))


```

## filter blast data by coordinates
```{r filter blast}
Sys.time()

library(GenomicRanges)
library(plyranges)

blast_rslt <- read.table("annotations/blastx.outfmt6", header = FALSE, sep = "\t")

colnames(blast_rslt) <- c("query", "sseqid", "pident", "length", "mismatch", "gapopen", 
                          "qstart", "qend", "sstart", "send", "evalue", "bitscore")

test <- blast_rslt %>% dplyr::select(query, sseqid, qstart, qend, evalue) 

test <- test %>% 
  mutate(i_start = case_when(qstart < qend ~ qstart,
                             qend < qstart ~ qend))

test <- test %>% 
  mutate(i_end = case_when(qstart < qend ~ qend,
                           qend < qstart ~ qstart))

test <- test %>% 
  mutate(strand = case_when(qstart < qend ~ "+", 
                            qend < qstart ~ "-"))

test <- test %>% 
  select(query, sseqid, i_start, i_end, evalue, strand)

colnames(test) <- c("query", "seqnames", "start", "end", "evalue", "strand")

test_irange <- test %>% 
  as_granges()

test_disjoin <- reduce(test_irange,with.revmap = TRUE)

list_revmap <- as.data.frame(mcols(test_disjoin))

filtered_data <- c()
for(i in 1:nrow(list_revmap)){
  filtered_data <- c(filtered_data, (slice(list_revmap, i) %>% 
                                       unlist(use.names=FALSE))[which.min(slice(test,  slice(list_revmap, i) %>%
                                                                                  unlist(use.names=FALSE))$evalue)])
}

best_hits <- slice(test, filtered_data)
```

## create metaphlan table at species level for arby to run through multimedia
```{r prep for multimedia}
Sys.time()

metaphlan <- readxl::read_excel("data/msystems.00452-24-s0004.xlsx")
colnames(metaphlan) <- gsub("-005S|-004S|-003S|-002S", "", colnames(metaphlan))
metaphlan <- metaphlan %>% 
  column_to_rownames("Taxonomy\\Sample ID")

# Define a function to create a table for a given taxonomic level
create_taxa_table <- function(metaphlan, level) {
  level_names <- c("domain", "phylum", "class", "order", "family", "genus", "species")
  level_index <- match(level, level_names)
  
  # Create a list of columns to remove based on the current level
  cols_to_remove <- level_names[(level_index + 1):length(level_names)]
  
  metaphlan %>%
    rownames_to_column() %>%
    separate(col = "rowname", sep = "\\|", into = level_names, fill = "right") %>%
    # Filter out rows where the next level is NA, but the current level is not NA
    filter(if (level_index < length(level_names)) is.na(get(level_names[level_index + 1])) & !is.na(get(level)) else !is.na(get(level))) %>%
    # Select the current level and the data columns
    select(all_of(c(level, setdiff(names(metaphlan), cols_to_remove)))) %>%
    # Set the current level as row names
    column_to_rownames(var = level)
}

# List of taxonomic levels
levels <- c("domain", "phylum", "class", "order", "family", "genus", "species")

# Apply the function to each level and store the results in a list
taxa_tables <- map(levels, ~create_taxa_table(metaphlan, .x))

# Set the names of the list elements to match the taxonomic levels
names(taxa_tables) <- levels

# Now you can access the phylum table using the name
species_table <- taxa_tables[["species"]]
write.csv(species_table, "tables/metaphlan_species.csv", quote = FALSE)
```




TODO: HIGH PRIORITY cluster analysis
TODO: account for repeated measure with diet as main effect and order of diet as repeated measure
TODO: WGCNA 
TODO: RLQ - which transcription factors controlling expression of each gene [Baty et al., 2013](https://pmc.ncbi.nlm.nih.gov/articles/PMC3686578/); [RLQ R tutorial](https://rfrelat.github.io/RLQ.html); [matt's google drive on rlq](https://drive.google.com/drive/folders/1BTs1d1G__SJJWVmdvMHpZJOLJmAZ16GO)


### END ###